tests/strategies/test_ddp_fully_sharded_native.py

```diff
@@ -0,0 +1,138 @@
+def test_fsdp_with_sharded_amp(device_count_mock, mock_cuda_available, tmpdir):
+    """Test to ensure that plugin native amp plugin is correctly chosen when using sharded."""
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, strategy="fsdp_native", gpus=1, precision=16)
+    assert isinstance(trainer.strategy, DDPFullyShardedNativeStrategy)
+    assert isinstance(trainer.strategy.precision_plugin, FullyShardedNativeMixedPrecisionPlugin)
+
+
+class TestFSDPModel(BoringModel):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.layer: Optional[torch.nn.Module] = None
+
+    def _init_model(self) -> None:
+        self.layer = torch.nn.Sequential(torch.nn.Linear(32, 32), torch.nn.ReLU(), torch.nn.Linear(32, 2))
+
+    def setup(self, stage: str) -> None:
+        if self.layer is None:
+            self._init_model()
+
+    def configure_sharded_model(self) -> None:
+        # the model is already wrapped with FSDP: no need to wrap again!
+        if isinstance(self.layer, FullyShardedDataParallel):
+            return
+        for i, layer in enumerate(self.layer):
+            if i % 2 == 0:
+                self.layer[i] = wrap(layer)
+        self.layer = wrap(self.layer)
+
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        # when loading full state dict, we first need to create a new unwrapped model
+        self._init_model()
+
+    def configure_optimizers(self):
+        return torch.optim.SGD(self.layer.parameters(), lr=0.1)
+
+    def on_train_start(self) -> None:
+        self._assert_layer_fsdp_instance()
+
+    def on_test_start(self) -> None:
+        self._assert_layer_fsdp_instance()
+
+    def on_validation_start(self) -> None:
+        self._assert_layer_fsdp_instance()
+
+    def on_prediction_start(self) -> None:
+        self._assert_layer_fsdp_instance()
+
+    def _assert_layer_fsdp_instance(self) -> None:
+        assert isinstance(self.layer, FullyShardedDataParallel)
+        assert isinstance(self.layer.module[0], FullyShardedDataParallel)
+        assert isinstance(self.layer.module[2], FullyShardedDataParallel)
+        # root should not be resharding
+        assert self.layer.reshard_after_forward is False
+        # Assert that the nested layers are set reshard_after_forward to True
+        assert self.layer.module[0].reshard_after_forward is True
+        assert self.layer.module[2].reshard_after_forward is True
+
+
+@RunIf(min_gpus=1, skip_windows=True, standalone=True)
+def test_fully_sharded_native_strategy_checkpoint(tmpdir):
+    """Test to ensure that checkpoint is saved correctly when using a single GPU, and all stages can be run."""
+
+    model = TestFSDPModel()
+    trainer = Trainer(default_root_dir=tmpdir, gpus=1, strategy="fsdp_native", precision=16, max_epochs=1)
+    _run_multiple_stages(trainer, model, os.path.join(tmpdir, "last.ckpt"))
+
+
+@RunIf(min_gpus=2, skip_windows=True, standalone=True)
```

Comment:
you can add `min_torch="1.11"` to make sure that these tests run when torch version 1.11 or more is available.